### PR TITLE
adds authentication

### PR DESCRIPTION
**Pull request by charlie-healthfinch**
# What

Implements a `get_user_authentication` method so clients can authenticate prior to making magic api calls.

# Why

Starting in January, it is expected that all users will authenticate via the `GetUserAuthentication` endpoint prior to making other magic api calls. This requires users of our library to do so.

# How

In keeping with our current convention, we

* take an EHR userid/password in the configuration of a client (via `AllscriptsUnityClient.create`), and
* require the user to explicitly ask for that authentication to happen (via `#get_user_authentication`).

Additionally, we now raise an error if the user tries to call any magic api methods without authenticating.

# Example

Examples of usage (without giving any usernames/passwords away)
```
# With the valid sandbox EHR username/password
[14] pry(main)> client = AllscriptsUnityClient.create(config_options.merge(valid_user_options));
[15] pry(main)> client.get_security_token!;
[16] pry(main)> client.get_user_authentication
=> true

# With an invalid EHR password for the sandbox
[18] pry(main)> client = AllscriptsUnityClient.create(config_options.merge(invalid_user_options));
[19] pry(main)> client.get_security_token!;
[20] pry(main)> client.get_user_authentication
=> false
```

### CODE REVIEW

File: spec/json_client_driver_spec.rb [hunk 1]
```diff
@@ -47,37 +47,50 @@
   end
 
   describe '#magic' do
-    before do
-      stub_request(:post, "http://www.example.com/Unity/UnityService.svc/json/MagicJson").
-               to_return(status: 200, body: get_server_info, headers: {})
-    end
+    context 'when the user has authenticated' do
```
Comment (charlie-healthfinch) [inline spec/json_client_driver_spec.rb]: I added a `context` here to distinguish authenticated v. unauthenticated calls, so the next 30 lines or so are mostly just indentation changes.

File: lib/allscripts_unity_client.rb [hunk 2]
```diff
@@ -21,6 +21,9 @@ class APIError < RuntimeError
   class GetSecurityTokenError < RuntimeError
   end
 
+  # Raised when user attempst to make unauthenticated calls.
```
Comment (futuro) [inline lib/allscripts_unity_client.rb]:
```suggestion
  # Raised when a user attempts to make unauthenticated calls.
```

File: README.md [hunk 3]
```diff
@@ -28,7 +28,9 @@ unity_client = AllscriptsUnityClient.create({
   base_unity_url: "http://unity.base.url",
   appname: "appname",
   username: "username",
-  password: "password"
+  password: "password",
```
Comment (futuro) [inline README.md]: @charlie-healthfinch mind if I push a commit to change the username and password values to `unity license username` and `unity license password`?

File: spec/json_client_driver_spec.rb (hunk 1 above)
Comment (futuro) [inline spec/json_client_driver_spec.rb]:
```suggestion
    context 'when the user is authenticated' do
```
?

**Comment by futuro** [comment]
This is quite nice, thank you! Before accepting, I want to make sure that the travis-ci builds failing isn't an issue.

File: README.md [hunk 4]
```diff
@@ -28,7 +28,9 @@ unity_client = AllscriptsUnityClient.create({
...
+  ehr_userid: "userid",
```
Comment (defndaines) [inline README.md]: As I was looking over things, I see that there are several methods which are currently expecting a `userid` to be passed. For example, `get_medication_by_trans_id`. I assume this is the same value. It suggests that we are going to need some refactoring around that, though that doesn't necessarily have to happen in this PR.

File: lib/allscripts_unity_client/json_client_driver.rb [hunk 5]
```diff
@@ -118,6 +141,10 @@ def retire_security_token!(parameters = {})
       @security_token = nil
     end
 
+    def user_authenticated?
+      @user_authentication.present?
```
Comment (defndaines) [inline lib/allscripts_unity_client/json_client_driver.rb]: Should we also nullify this value if the `retire_security_token!` method is called?

File: spec/json_client_driver_spec.rb [hunk 6]
```diff
@@ -149,4 +162,49 @@
       it { expect { subject.send(:raise_if_response_error, error_string) }.to raise_error(AllscriptsUnityClient::APIError) }
     end
   end
+
+  context 'user authentication' do
+    context 'when the user is attempting to authenticate' do
+      let(:userid) { 'user1' }
+      let(:password) { 'password' }
```
Comment (defndaines) [inline spec/json_client_driver_spec.rb]: `p455w0rd` ? :)

File: README.md (hunk 3 above)
Comment (charlie-healthfinch) [reply README.md]: Works for me!

**Comment by charlie-healthfinch** [comment]
@futuro I'll push the changes in the other PR on top of this so we can have confidence in the CI tests.

File: README.md (hunk 3 above)
Comment (futuro) [reply README.md]: Great; done!

File: README.md (hunk 4 above)
Comment (charlie-healthfinch) [reply README.md]: Yea that's something we should probably clean up in the future.

**Comment by charlie-healthfinch** [comment]
I fixed the CI issues and we are all green here.

**Pull request merged**